src/librustc_typeck/collect.rs

```diff
@@ -968,17 +967,28 @@ fn generics_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
         });
     }
 
-    let type_param_to_index = types.iter()
-                                   .map(|param| (param.def_id, param.index))
-                                   .collect();
+    let opt_self = opt_self.into_iter().map(|ty| ty::GenericParamDef::Type(ty));
+    let lifetimes = regions.into_iter().map(|lt| ty::GenericParamDef::Lifetime(lt));
+    let types = types.into_iter().map(|ty| ty::GenericParamDef::Type(ty));
+    let params: Vec<_> = opt_self.chain(lifetimes)
+                                 .chain(types)
+                                 .collect();
+
+    let param_def_id_to_index =
+        params.iter()
+              .map(|param| {
+                  match param {
+                      ty::GenericParamDef::Lifetime(lt) => (lt.def_id, lt.index),
+                      ty::GenericParamDef::Type(ty) => (ty.def_id, ty.index),
```

Comment:
Aren't there `def_id` / `index` methods? Also, how about making `GenericParamDef` a `struct` with the common fields, plus an `enum` for lifetime vs type?

Reply:
And then making anything that currently takes a `RegionParamDef`/`TypeParamDef` take a `GenericParamDef` instead? Or extracting the common fields into another type, and then passing that and the specific kind def as a pair or something like that?